src/envoy/common/per_route_http_filter.h

```diff
@@ -0,0 +1,170 @@
+/* Copyright 2017 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef PROXY_PER_ROUTE_FILTER_H
+#define PROXY_PER_ROUTE_FILTER_H
+
+#include "envoy/http/filter.h"
+#include "envoy/router/router.h"
+
+#include <functional>
+#include <memory>
+
+namespace Envoy {
+namespace Http {
+
+template <typename InnerFilter>
+class PerRouteFilter : public Http::StreamFilter, public AccessLog::Instance {
+ public:
+  typedef std::unique_ptr<InnerFilter> InnerFilterPtr;
+  typedef std::function<InnerFilterPtr(const Envoy::Router::RouteEntry *)>
+      Constructor;
+
+  PerRouteFilter(Constructor constructor)
+      : constructor_(constructor), inner_filter_(nullptr) {}
+
+  void log(const Http::HeaderMap *request_headers,
+           const Http::HeaderMap *response_headers,
+           const RequestInfo::RequestInfo &request_info) override {
+    if (constructor_) {
+      Constructor constructor;
+      constructor.swap(constructor_);
+      inner_filter_ = constructor(request_info.routeEntry());
+    }
+    auto filter = dynamic_cast<AccessLog::Instance *>(inner_filter_.get());
+    if (filter) {
+      filter->log(request_headers, response_headers, request_info);
+    }
+  }
+
+  void onDestroy() override {
+    auto filter = dynamic_cast<Http::StreamFilterBase *>(inner_filter_.get());
```

Comment:
Is it necessary to use dynamic_cast in every function? Can we use inner_filter_->onDestroy() instead? I think it is recommended to avoid use too many dynamic_cast in production code [[link](https://google.github.io/styleguide/cppguide.html#Run-Time_Type_Information__RTTI_)].

Reply:
Not really, a inner_filter_ might not be a instance of StreamFilter when it is only a AccessLog::instance.

Ideally I can avoid all dynamic_cast in this class with a heavy template programming, but I didn't want to go that far at this moment.